ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryLimitFunctionalTest.java

```diff
@@ -0,0 +1,491 @@
+        final List<StreamedRow> rows_3 = makePullQueryRequest(clusterFormation.router.getApp(),
+                sqlLimit, null, USER_CREDS);
+
+        // check that we only got limit number of rows
+        assertThat(rows_3, hasSize(limitSubsetSize));
+    }
+
+    @Test
+    public void shouldReturnLimitRowsMultiHostSetupStream() {
+        // Given:
+        final int numLimitRows = 200;
+        final int limitSubsetSize = HEADER + numLimitRows + LIMIT_REACHED_MESSAGE;
+        ClusterFormation clusterFormation = findClusterFormation(TEST_APP_0, TEST_APP_1, TEST_APP_2);
+        waitForClusterToBeDiscovered(clusterFormation.router.getApp(), 3, USER_CREDS);
+        waitForRemoteServerToChangeStatus(clusterFormation.router.getApp(),
+                clusterFormation.router.getHost(), HighAvailabilityTestUtil.lagsReported(3), USER_CREDS);
+
+
+        final String sqlStreamPull = "SELECT * FROM " + USERS_STREAM + ";";
+        final String sqlLimit = "SELECT * FROM " + USERS_STREAM + " LIMIT " + numLimitRows + " ;";
+
+        //scan the entire stream first
+        final List<StreamedRow> rows_0 = makePullQueryRequest(clusterFormation.router.getApp(),
+                sqlStreamPull, null, USER_CREDS);
+
+        //check that we got back all the rows
+        assertThat(rows_0, hasSize(HEADER + TOTAL_RECORDS));
+
+        //issue pull query on stream with limit
+        final List<StreamedRow> rows_1 = makePullQueryRequest(clusterFormation.router.getApp(),
+                sqlLimit, null, USER_CREDS);
+
+        // check that we only got limit number of rows
+        assertThat(rows_1, hasSize(limitSubsetSize));
+
+        // Partition off the active
+        clusterFormation.active.getShutoffs().shutOffAll();
+
+        waitForRemoteServerToChangeStatus(
+                clusterFormation.router.getApp(),
+                clusterFormation.standBy.getHost(),
+                HighAvailabilityTestUtil::remoteServerIsUp,
+                USER_CREDS);
+        waitForRemoteServerToChangeStatus(
+                clusterFormation.router.getApp(),
+                clusterFormation.active.getHost(),
+                HighAvailabilityTestUtil::remoteServerIsDown,
+                USER_CREDS);
+
+
+        // issue stream scan after partitioning active
+        final List<StreamedRow> rows_2 = makePullQueryRequest(clusterFormation.router.getApp(),
+                sqlStreamPull, null, USER_CREDS);
+
+        // check that we get all rows back
+        assertThat(rows_2, hasSize(HEADER + TOTAL_RECORDS));
+        assertThat(rows_0, is(matchersRowsAnyOrder(rows_2)));
+
+        // issue pull query with limit after partitioning active
+        final List<StreamedRow> rows_3 = makePullQueryRequest(clusterFormation.router.getApp(),
+                sqlLimit, null, USER_CREDS);
+
+        // check that we only got limit number of rows
+        assertThat(rows_3, hasSize(limitSubsetSize));
+    }
+
+    private ClusterFormation findClusterFormation(
```

Comment:
I don't really think that most of this is necessary for your test.  I guess I understand a simplified version of this where you ensure that you have an active host for all of your partitions.